test/Microsoft.IdentityModel.Xml.Tests/EnvelopedSignatureWriterTests.cs

```diff
@@ -135,6 +138,277 @@ public static TheoryData<EnvelopedSignatureTheoryData> RoundTripSaml2TheoryData
+                        {
+                            esw.WriteStartElement("EntityDescriptor", "urn:oasis:names:tc:SAML:2.0:metadata");
+                            esw.WriteSignature();
+                            esw.WriteAttributeString("entityID", "issuer");
+                            esw.WriteStartElement("RoleDescriptor", "urn:oasis:names:tc:SAML:2.0:metadata");
+                            esw.WriteEndElement();
+                            esw.WriteEndElement();
+                        },
+                        ReferenceId = Default.ReferenceUriWithOutPrefix,
+                        SigningCredentials = Default.AsymmetricSigningCredentials,
+                        ExpectedException = new ExpectedException(typeof(System.Reflection.TargetInvocationException), null, typeof(System.InvalidOperationException)),
+                        TestId = "WriteSignatureInTheMiddleBeforeAttributeInvalid",
+                    }
+                };
+            }
+        }
+
+        [Fact]
+        public void RoundTripSamlP()
+        {
+            var context = new CompareContext($"{this}.RoundTripSamlP");
+            ExpectedException expectedException = ExpectedException.NoExceptionExpected;
+            var samlpTokenKey = KeyingMaterial.RsaSigningCreds_4096_Public.Key;
+            var samlpTokenSigningCredentials = KeyingMaterial.RsaSigningCreds_4096;
+            var samlpKey = KeyingMaterial.RsaSigningCreds_2048_Public.Key;
+            var samlpSigningCredentials = KeyingMaterial.RsaSigningCreds_2048;
+
+            try
+            {
+                // write samlp
+                var settings = new XmlWriterSettings
+                {
+                    Encoding = new UTF8Encoding(false)
+                };
+                var buffer = new MemoryStream();
+                var esw = new EnvelopedSignatureWriter(XmlWriter.Create(buffer, settings), samlpSigningCredentials, "id-uAOhNLe7abGB6WGPk");
+
+                esw.WriteStartElement("ns0", "Response", "urn:oasis:names:tc:SAML:2.0:protocol");
+                esw.WriteAttributeString("ns1", "urn:oasis:names:tc:SAML:2.0:assertion");
+                esw.WriteAttributeString("ns2", "http://www.w3.org/2000/09/xmldsig#");
+                esw.WriteAttributeString("Destination", "https://tnia.eidentita.cz/fpsts/processRequest.aspx");
+                esw.WriteAttributeString("ID", "id-uAOhNLe7abGB6WGPk");
+                esw.WriteAttributeString("InResponseTo", "ida5714d006fcc430c92aacf34ab30b166");
+                esw.WriteAttributeString("IssueInstant", "2019-04-08T10:30:49Z");
+                esw.WriteAttributeString("Version", "2.0");
+                esw.WriteStartElement("ns1", "Issuer");
+                esw.WriteAttributeString("Format", "urn:oasis:names:tc:SAML:2.0:nameid-format:entity");
+                esw.WriteString("https://mojeid.regtest.nic.cz/saml/idp.xml");
+                esw.WriteEndElement();
+                esw.WriteSignature();
```

Comment:
If the SamlToken is written before this WriteSignature, the signatures will not validate.
We will need some logic to 'know' which signature is significant.

Message: Microsoft.IdentityModel.TestUtils.TestException : Microsoft.IdentityModel.Tokens.Xml.Tests.EnvelopedSignatureWriterTests.RoundTripSamlP


exception != null, expectedException.TypeExpected == null.
exception: Microsoft.IdentityModel.Xml.XmlValidationException: IDX30201: The 'Reference' did not validate: '#id-uAOhNLe7abGB6WGPk'.

Reply:
@brentschmaltz - Writing the signature explicitly (using WriteSignature) or using the default behavior (adding the signature as the last element) didn't change the digest and signature values in my tests.
I guess that the problem lies somewhere in EnvelopeSignatureReader/XmlTokenStreamReader.

When adding a signature after an assertion, an outer signature doesn't get recorded and therefore remains included in digested canonical bytes, leading to the mismatch in digest values.

Reply:
Yeah essentially each envelope signature writer needs to recognize it's 'signature'. One way may be to use a freshly generated Guid to mark the spot rather than a shared value.

Reply:
To obtain this exception, I moved the writing of the Saml2Assertion before the call to esw.WriteSignature as in:
````
esw.WriteEndElement();
Saml2Serializer samlSerializer = new Saml2Serializer(); 
Saml2Assertion assertion = CreateAssertion(samlpTokenSigningCredentials); 
samlSerializer.WriteAssertion(esw, assertion); 
esw.WriteSignature();
````

Reply:
Addressed this issue by slightly altering the logic in EnvelopedSignatureReader and XmlTokenStreamReader classes.